src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java

```diff
@@ -773,6 +773,88 @@ public void test100PercentContaminationNoCallsInGVCFMode() throws Exception {
         }
     }
 
+    @DataProvider
+    public Object[][] getMaxAlternateAllelesData() {
+        return new Object[][] {
+                // bam, reference, interval string, max alternate alleles, GVCF mode toggle
+                { NA12878_20_21_WGS_bam, b37_reference_20_21, "20:10000000-10010000", 1, false },
+                { NA12878_20_21_WGS_bam, b37_reference_20_21, "20:10000000-10010000", 2, true }
+        };
+    }
+
+    /*
+     * Test for the --max-alternate-alleles argument
+     */
+    @Test(dataProvider = "getMaxAlternateAllelesData")
+    public void testMaxAlternateAlleles(final String bam, final String reference, final String intervalString,
+                                        final int maxAlternateAlleles, final boolean gvcfMode) {
+        final File outputNoMaxAlternateAlleles = createTempFile("testMaxAlternateAllelesNoMaxAlternateAlleles", (gvcfMode ? ".g.vcf" : ".vcf"));
+        final File outputWithMaxAlternateAlleles = createTempFile("testMaxAlternateAllelesWithMaxAlternateAlleles", (gvcfMode ? ".g.vcf" : ".vcf"));
+
+        // Run both with and without --max-alternate-alleles over our interval, so that we can
+        // prove that the argument is working as intended.
+        final String[] argsNoMaxAlternateAlleles = {
+                "-I", bam,
+                "-R", reference,
+                "-L", intervalString,
+                "-O", outputNoMaxAlternateAlleles.getAbsolutePath(),
+                "-ERC", (gvcfMode ? "GVCF" : "NONE")
+        };
+        runCommandLine(argsNoMaxAlternateAlleles);
+
+        final String[] argsWithMaxAlternateAlleles = {
+                "-I", bam,
+                "-R", reference,
+                "-L", intervalString,
+                "-O", outputWithMaxAlternateAlleles.getAbsolutePath(),
+                "--max-alternate-alleles", Integer.toString(maxAlternateAlleles),
+                "-ERC", (gvcfMode ? "GVCF" : "NONE")
+        };
+        runCommandLine(argsWithMaxAlternateAlleles);
+
+        final List<VariantContext> callsNoMaxAlternateAlleles = VariantContextTestUtils.readEntireVCFIntoMemory(outputNoMaxAlternateAlleles.getAbsolutePath()).getRight();
+        final List<VariantContext> callsWithMaxAlternateAlleles = VariantContextTestUtils.readEntireVCFIntoMemory(outputWithMaxAlternateAlleles.getAbsolutePath()).getRight();
+
+        // First, assert that there is at least one call in the VCF produced WITHOUT --max-alternate-alleles
+        // that has more than maxAlternateAlleles alt alleles, excluding NON_REF (otherwise the test below
+        // won't be meaningful).
+        boolean sawTooManyAllelesWithoutMaxAlternateArg = false;
+        for ( final VariantContext vc : callsNoMaxAlternateAlleles ) {
+            if ( getNumAltAllelesExcludingNonRef(vc) > maxAlternateAlleles ) {
+                sawTooManyAllelesWithoutMaxAlternateArg = true;
+            }
+        }
+        Assert.assertTrue(sawTooManyAllelesWithoutMaxAlternateArg,
```

Comment:
Ooh we should really do this more often.  I bet there are old integration tests where the data was changed and we lost coverage.

Reply:
It's a tradeoff -- it provides continuous proof that the test is meaningful, but at the expense of doubling the runtime of the test.

Reply:
Actually, if you're really concerned about doubling the runtime, you can merge those two loops since there should be the same number of VCs in the same order between the two files, right?  And if they don't line up then we're in big big trouble anyway.

Reply:
You have to actually run the `HaplotypeCaller` twice to generate the two VCFs in the first place (with and without the argument being tested). Notice the two calls to `runCommandLine()` above -- that's what doubles the runtime.